ledger/src/blockstore.rs

```diff
@@ -2490,7 +2490,22 @@ impl Blockstore {
     }
 
     pub fn get_index(&self, slot: Slot) -> Result<Option<Index>> {
-        self.index_cf.get(slot)
+        // Migration strategy for new column format:
+        // 1. Release 1: Add ability to read new format as fallback, keep writing old format
+        // 2. Release 2: Switch to writing new format, keep reading old format as fallback
+        // 3. Release 3: Remove old format support once stable
+        // This allows safe downgrade to Release 1 since it can read both formats
+        // https://github.com/anza-xyz/agave/issues/3570
+        self.index_cf.get_with(slot, |slice| {
+            let index: bincode::Result<Index> = bincode::deserialize(slice);
+            match index {
+                Ok(index) => Ok(index),
+                Err(_) => {
```

Comment:
what guarantees that bincode deserialize here fails if the serialized bytes were actualy `IndexV2`?

Reply:
So my question is that what guarantees that bytes obtained from serializing an `IndexV2` do not deserialize into a "bogus" `BTreeSet<u64>`?

IOW, are the set of all byte arrays obtained from serializing an `IndexV2` disjoint from the set of byte arrays that can be deserialized into a `BTreeSet<u64>`?

Reply:
`serde_bytes::Bytes` used here uses `Serializer::serialize_bytes`:
https://github.com/serde-rs/bytes/blob/07f078abd/src/ser.rs#L50
which is len prefixed:
https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_bytes

and even if you were serializing an array (ie. without len prefixed), you still can serialize an array which deserializes into a valid sequence (including BTreeSet):

```rust
#[test]
fn test_btree_serde() {
    let arr: [u64; 2] = [1u64, 2];
    let bytes = bincode::serialize(&arr).unwrap();
    assert_eq!(bytes.len(), 2 * 8);
    let out: BTreeSet<u64> = bincode::deserialize(&bytes).unwrap();
    dbg!(&out);
}
```

Reply:
> which is actually implemented as [a tuple](https://docs.rs/serde/1.0.215/src/serde/ser/impls.rs.html#143-164) under the hood, making it classified as a collection (length prefixed).

not true, there is no length prefix.
see `assert_eq!(bytes.len(), 2 * 8)` or just add `dbg!(&bytes)`.

Your implementation on the other hand is doing length prefixed because of using  `serialize_bytes`.

Also this part of comments earlier were wrong too:
> the first 4 bytes will be read as len
> it will have 4 extra bytes (the len).

bincode serializes length as u64. so 8 bytes with fixint encoding.

Reply:
> bincode serializes length as u64. so 8 bytes with fixint encoding.

My bad, I mixed up enum discriminants and collection lengths

Zach's point remains valid: `BTreeSet<u64>` can never be deserialized from any bit pattern of `ShredIndexV2`. That's because `ShredIndexV2` has an hardcoded length of >= 4096 (> if we ever expand it), but those are 4096 _bytes_ not _u64_. So deserializing `BTreeSet<u64>` from a serialized `ShredIndexV2` will always run out of data, since the latter serialized is 8 + 4096 bytes while deserializing `BTreeSet<u64>` is going to want to read 8 + 4096 * 8 bytes.

We should probably put this reasoning as a comment somewhere so we don't get confused again 2 weeks from now :D

Reply:
There is more to it here because the code here is not de-serializing bytes obtained just from:
```rust
type ShredIndexV2Inner = [u64; NUM_U64_PER_SHRED_INDEX_V2];
```
but it is deserializing bytes obtained from:
```rust
pub struct IndexV2 {
    pub slot: Slot,
    data: ShredIndexV2,
    coding: ShredIndexV2,
}
``` 
into an `Index` which has `BtreeSet<u64>` fields.

So there are a lot of trailing bytes for the first `ShredIndexV2Inner` and by the time you deserialize the first `BtreeSet<u64>`, you might be at an offset that the 2nd `BtreeSet<u64>` be deserialized as well because you read a length (at a wrong offset) which is not the same as the 2nd `ShredIndexV2Inner`.

That being said, I _think_ you can still show that the trailing bytes here are not sufficient for `bincode::deserialize::<Index>` to work.

The other issue here is that to make sure that this PR (serde code or elsewhere) does not tie our hands with respect to changing (increasing or decreasing) `MAX_DATA_SHREDS_PER_SLOT`.
We had outages which forced us to "decrease" `MAX_CODE_SHREDS_PER_SLOT`: https://github.com/solana-labs/solana/pull/30543 and I cannot rule out the same for `MAX_DATA_SHREDS_PER_SLOT`.
Can you confirm if we change (again either increase or decrease) `MAX_DATA_SHREDS_PER_SLOT` then nothing in this PR will break (serde code or elsewhere)?
Note that we might have to run the old code on a rocksdb generated by the new code (e.g. upgrade, run a node for a little while, then downgrade).